### PR TITLE
support --ignored-image-list ignoredlist args

### LINKED ISSUE #1553: when sealer build save domain docker.io image error: get repository name error: invalid reference format
#### What happened:
我在尝试Build application image Kubevela时出现了以下报错
imageList

```
oamdev/alpine-k8s:1.18.2
oamdev/cluster-gateway:v1.4.0
oamdev/hello-world:v1
oamdev/kube-webhook-certgen:v2.4.1
oamdev/vela-core:v1.4.5
```

build执行
```
# sealer build -t kubevela-app:v1.4.5 --base=false  .
```

### Relevant log output?
出现报错如下：
```Shell
Error: save domain docker.io image error: get repository name error: invalid reference format
Usage:
  sealer build [flags] PATH

Examples:
the current path is the context path, default build type is lite and use build cache

build:
        sealer build -f Kubefile -t my-kubernetes:1.19.8 .

build without cache:
        sealer build -f Kubefile -t my-kubernetes:1.19.8 --no-cache .

build without base:
        sealer build -f Kubefile -t my-kubernetes:1.19.8 --base=false .

build with args:
        sealer build -f Kubefile -t my-kubernetes:1.19.8 --build-arg MY_ARG=abc,PASSWORD=Sealer123 .



Flags:
      --base                build with base image,default value is true. (default true)
      --build-arg strings   set custom build args
  -h, --help                help for build
  -t, --imageName string    cluster image name
  -f, --kubefile string     kubefile filepath (default "Kubefile")
  -m, --mode string         cluster image build type, default is lite (default "lite")
      --no-cache            build without cache
      --platform string     set cloud image platform,if not set,keep same platform with runtime

Global Flags:
      --config string   config file (default is $HOME/.sealer.json)
  -d, --debug           turn on debug mode
      --hide-path       hide the log path
      --hide-time       hide the log time

save domain docker.io image error: get repository name error: invalid reference format
```

debug时候偶现pull docker.io/library/_params.image:latest时的报错：
```
save domain docker.io image error: get library/parameter.image tag descriptor error: unknown tag=latest, try "docker login" if you are using a private registry
```

### How to reproduce it (as minimally and precisely as possible)?

```YAML
Kubefile

FROM kubernetes:v1.22.8

#images pull before cluster build
COPY imageList manifests
#helm

# install kubevela command
COPY kubectl-vela-v1.4.5-linux-amd64.tar.gz .
CMD tar zxvf kubectl-vela-v1.4.5-linux-amd64.tar.gz && cp linux-amd64/kubectl-vela /usr/local/bin/vela && chmod +x /usr/local/bin/vela 

# install kubevela core
COPY vela-core charts
CMD  helm install --create-namespace -n vela-system kubevela charts/vela-core --wait

# install velaUX
CMD /usr/local/bin/vela addon enable velaux --version=v1.4.5 serviceType=NodePort
```

### What is the version of Sealer you using?

{"gitVersion":"v0.8.5","gitCommit":"f9c3d99","buildDate":"2022-04-28 14:16:58","goVersion":"go1.16.15","compiler":"gc","platform":"linux/amd64"}

### What is your OS environment?

CentOS Linux 7 (Core)

### What is the Kernel version?

3.10.0-957.el7.x86_64 #1 SMP Thu Nov 8 23:39:32 UTC 2018 x86_64 x86_64 x86_64 GNU/Linux

### Other environment you want to tell us?

- Cloud provider or hardware configuration:
- Install tools:
- Others:

### LINKED ISSUE #1913: An error occurred at   sealer build
### What happen?

我在用sealer 制作基础镜像时在里面放了一个chart包(charts/vela-core)，然后build时解析这个包的镜像时有问题，解析出一个错误的包
docker.io/library/parameter.image:latest
docker.io/library/parameter.image:latest
实际上没有这些包

### Relevant log output?

```Shell
[root@master01 builad]# sealer  build  -t  testtest:1.0.2  . 
2022-12-08 18:05:04 [INFO] [utils.go:135] get mount dir :/var/lib/sealer/tmp/.DTmp-061589638 success 
2022-12-08 18:05:04 [INFO] [executor.go:123] start to check the middleware file
2022-12-08 18:05:04 [INFO] [executor.go:63] run build layer: COPY . .
2022-12-08 18:05:05 [INFO] [executor.go:95] exec all build instructs success
Pulling image: docker.io/kubernetesui/dashboard:v2.6.1
Pulling image: docker.io/kubernetesui/metrics-scraper:v1.0.8
Pulling image: docker.io/library/parameter.image:latest
Pulling image: docker.io/library/string:latest
Pulling image: docker.io/library/_params.image:latest
Pulling image: docker.io/oamdev/vela-core:v1.6.4
Pulling image: docker.io/oamdev/kube-webhook-certgen:v2.4.1
```

### What you expected to happen?

后期能不能加一些选项，比如忽略一些错误的镜像

### How to reproduce it (as minimally and precisely as possible)?

```YAML
FROM kubernetes:v1.24.6
COPY . .
CMD cp -ar bin/helm   /usr/local/bin/ && cp -ar bin/vela   /usr/local/bin/

CMD helm install --create-namespace -n vela-system kubevela charts/vela-core --wait  &&  vela addon enable velaux
CMD  kubectl  apply -f  manifests/recommended.yaml
```

### Anything else we need to know?

_No response_

### What is the version of Sealer you using?

{"gitVersion":"v0.8.6","gitCommit":"884513e","buildDate":"2022-07-12 02:58:54","goVersion":"go1.16.15","compiler":"gc","platform":"linux/amd64"}

### What is your OS environment?

CentOS Linux 7 (Core)

### What is the Kernel version?

Linux master01 3.10.0-1160.76.1.el7.x86_64 #1 SMP Wed Aug 10 16:21:17 UTC 2022 x86_64 x86_64 x86_64 GNU/Linux

### Other environment you want to tell us?

- Cloud provider or hardware configuration:
- Install tools:
- Others:

### LINKED ISSUE #2001: [Task List] 0.9.2 Release - 20230228
## Issue Description

Type: *feature request*

### Universal capability
- [x] support multi-arch for sealer image
- [ ] support e2e test #1966
   - [x] support e2e test CI
   - [ ] test coverage reaches 30 percent
   - [x] support running e2e test locally
- [ ] Performance, disk usage optimization
   - [ ]  use image cache when build stage

### APP Image Related
- [x] support image policy with Kyverno
- [x]   continue to improve APP,APPCMDS instruction for Kubefile and related config for Clusterfile
- [ ] completely decouple the app image from the cluster image
  - [ ] run app image on any declared supported versions of the kubernetes cluster 
- [x] support for concurrent deployment of applications 
- [x] https://github.com/sealerio/sealer/issues/2017
### Cluster Image Related
- [x] support upgrade kubernetes cluster
- [x] support Linux user with root privileges to install ClusterImage  #1999 
- [x] support more container runtime
  - [x] support containerd
- [x] support more cluster runtime
  - [x] support k0s cluster runtime

### Bugfix
- [x] https://github.com/sealerio/sealer/issues/2000 @Stevent-fei 
- [x] https://github.com/sealerio/sealer/issues/1998 @Stevent-fei 
- [x] https://github.com/sealerio/sealer/issues/1995 @Starnop 
- [x] https://github.com/sealerio/sealer/issues/1994 @Stevent-fei
- [x] https://github.com/sealerio/sealer/issues/1993 @kakaZhou719 
- [x] https://github.com/sealerio/sealer/issues/1989 @kakaZhou719 
- [x] https://github.com/sealerio/sealer/issues/1986 @kakaZhou719

### LINKED ISSUE #2007: [FEATURE] add ignore image list flag for supporting user to choose which one will be saved
## Issue Description

Type: *feature request*

add flag `--ignore-image-list` to sealer build : `sealer build --ignored-image-list ignoredlist`

write ignored file:

```shell
[root@iZbp1daq0hry4n4gd6lelwZ]# cat ignoredlist 
mysql:3.2
docker.io/library/nginx:alpine
```

when build : sealer will not save those images read from  ignorelist file.

### PR DESCRIPTION
**Pull request by Stevent-fei**
### Describe what this PR does / why we need it

support sealer build --ignored-image-list ignoredlist args

example:

sealer build -f Kubefile -t docker.io/sealerio/kubernetes:v1.22.15 --ignored-image-list ignoredlist

### Does this pull request fix one issue?

Fixes [#1994](https://github.com/sealerio/sealer/issues/1994)
Fixes #2007
Fixes #1913
Fixes #1553
Fixes #2001

### CODE REVIEW

File: cmd/sealer/cmd/utils/image.go [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+	for v := range set2 {
+		if !set1[v] {
+			b = append(b, v)
+		}
+	}
+	return append(a, b...)
+}
+
+// Intersection get the intersection elements from the two arrays to form a new array
+func Intersection(nums1 []string, nums2 []string) (intersection []string) {
```
Comment (kakaZhou719) [inline cmd/sealer/cmd/utils/image.go:41]: we dont need this ,using utils/strings

File: cmd/sealer/cmd/utils/image.go [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package utils
+
+// FindDifference get different elements from two arrays to form a new array
+func FindDifference(nums1, nums2 []string) []string {
```
Comment (kakaZhou719) [inline cmd/sealer/cmd/utils/image.go:18]: same

File: build/layerutils/manifests/manifests.go [hunk 3]
```diff
@@ -35,13 +36,14 @@ func (manifests *Manifests) ListImages(yamlFile string) ([]string, error) {
 
 	images := layerutils.DecodeImages(string(yamlBytes))
 	if len(images) != 0 {
-		list = append(list, images...)
+		for _, image := range images {
+			named, err := reference.ParseToNamed(image)
```
Comment (kakaZhou719) [inline build/layerutils/manifests/manifests.go:40]: 1. no need return err, just continue 
2. image validation should be applied for all image list searched from build context.

File: build/buildimage/differ.go [hunk 4]
```diff
@@ -118,7 +119,17 @@ func ParseContainerImageList(srcPath string) ([]*v12.ContainerImage, error) {
...
-			containerImageList = append(containerImageList, ima...)
+			for _, image := range ima {
+				img, err := reference2.ParseNormalizedNamed(image.Image)
+				if err != nil {
+					continue
+				}
+				containerImageList = append(containerImageList, &v12.ContainerImage{
+					Image:    img.Name(),
+					AppName:  "",
+					Platform: nil,
```
Comment (kakaZhou719) [inline build/buildimage/differ.go:130]: Platform , AppName could not be nil

File: cmd/sealer/cmd/image/build.go [hunk 5]
```diff
@@ -332,36 +337,47 @@ func applyRegistryToImage(engine imageengine.Interface, imageID string, platform
...
-		containerImageList = append(containerImageList, containerImage)
+
+	// ignored image list
+	if buildFlags.IgnoredImageList != "" && osi.IsFileExist(buildFlags.IgnoredImageList) {
+		ignoredImages, err := osi.NewFileReader(buildFlags.IgnoredImageList).ReadLines()
+		if err != nil {
+			return "", nil, err
+		}
+
+		containerImageList = append(containerImageList, parsedContainerImageList...)
```
Comment (kakaZhou719) [inline cmd/sealer/cmd/image/build.go:355]: bug: we should append two containerImageList out of this logic

File: cmd/sealer/cmd/image/build.go [hunk 6]
```diff
@@ -315,8 +322,6 @@ func applyRegistryToImage(engine imageengine.Interface, imageID string, platform
 		return "", nil, errors.Wrapf(err, "failed to create working container, imageid: %s", imageID)
 	}
 
-	var containerImageList []*v12.ContainerImage
-
 	// download container image from `imageList`
 	if buildFlags.ImageList != "" && osi.IsFileExist(buildFlags.ImageList) {
 		images, err := osi.NewFileReader(buildFlags.ImageList).ReadLines()
```
Comment (kakaZhou719) [inline cmd/sealer/cmd/image/build.go:327]: add  the validation of the image format for image-list

Comment (Stevent-fei) [reply cmd/sealer/cmd/image/build.go:276]: > add the validation of the image format for image-list

added

File: cmd/sealer/cmd/image/build.go [hunk 7]
```diff
@@ -63,7 +63,9 @@ build with image type:
   sealer build -f Kubefile -t my-kubernetes:1.19.8 --type=kube-installer(default)
   app-installer type image will not install kubernetes.
 build multi-platform image:
-	sealer build -f Kubefile -t my-kubernetes:1.19.8 --platform linux/amd64,linux/arm64
+  sealer build -f Kubefile -t my-kubernetes:1.19.8 --platform linux/amd64,linux/arm64
+build manually ignore image:
+  sealer build -f Kubefile -t my-kubernetes:1.19.8 --ignored-image-list ignoredList
```
Comment (Starnop) [inline cmd/sealer/cmd/image/build.go:68]: s/ignoredList/ignoredListPath

File: cmd/sealer/cmd/image/build.go [hunk 8]
```diff
@@ -107,6 +109,7 @@ func NewBuildCmd() *cobra.Command {
 	//todo we can support imageList Flag to download extra container image rather than copy it to rootfs
 	buildCmd.Flags().StringVar(&buildFlags.ImageList, "image-list", "filepath", "`pathname` of imageList filepath, if set, sealer will read its content and download extra container")
 	buildCmd.Flags().StringVar(&buildFlags.ImageListWithAuth, "image-list-with-auth", "", "`pathname` of imageListWithAuth.yaml filepath, if set, sealer will read its content and download extra container images to rootfs(not usually used)")
+	buildCmd.Flags().StringVar(&buildFlags.IgnoredImageList, "ignored-image-list", "filepath", "`pathname` of ignored image list filepath, if set, sealer will read its contents and prevent downloading of the corresponding container")
```
Comment (Starnop) [inline cmd/sealer/cmd/image/build.go:112]: the corresponding container/ the corresponding container image

File: cmd/sealer/cmd/image/build.go [hunk 9]
```diff
@@ -239,6 +242,11 @@ func applyImagesWithSinglePlatform(engine imageengine.Interface, imageID string,
 }
 
 func applyRegistryToImage(engine imageengine.Interface, imageID string, platform v1.Platform) (string, []*v12.ContainerImage, error) {
+	var (
+		allContainerImageList []*v12.ContainerImage
```
Comment (Starnop) [inline cmd/sealer/cmd/image/build.go:246]: optimize the naming

File: cmd/sealer/cmd/image/build.go [hunk 10]
```diff
@@ -274,42 +280,46 @@ func applyRegistryToImage(engine imageengine.Interface, imageID string, platform
...
+				return "", nil, err
+			}
 			containerImageList = append(containerImageList, &v12.ContainerImage{
-				Image:    image,
+				Image:    img.String(),
 				AppName:  "",
 				Platform: &platform,
 			})
 		}
-		formatImages := buildimage.FormatImages(images)
```
Comment (Starnop) [inline cmd/sealer/cmd/image/build.go:283]: please make sure it's right to delete the related logic

Comment (Starnop) [reply cmd/sealer/cmd/image/build.go:283]: add e2e test here

File: build/buildimage/differ.go [hunk 11]
```diff
@@ -118,7 +119,17 @@ func ParseContainerImageList(srcPath string) ([]*v12.ContainerImage, error) {
 			if err != nil {
 				return fmt.Errorf("failed to parse images from %s: %v", dispatchType, err)
 			}
-			containerImageList = append(containerImageList, ima...)
+			for _, image := range ima {
```
Comment (Starnop) [inline build/buildimage/differ.go:122]: s/ima/imageList

**Review by kakaZhou719 — APPROVED** [review]
lgtm

**Pull request merged**